### PR TITLE
R2N

### PR DESCRIPTION
**Pull request by MohamedLaghdafHABIBOULLAH**

### CODE REVIEW

File: src/R2N.jl [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+A regularized quasi-Newton method for the problem
+
+    min f(x) + h(x)
+
+where f: ℝⁿ → ℝ has a Lipschitz-continuous Jacobian, and h: ℝⁿ → ℝ is
+lower semi-continuous and proper.
+
+About each iterate xₖ, a step sₖ is computed as an approximate solution of
+
+    min  φ(s; xₖ) + ½ σₖ ‖s‖² + ψ(s; xₖ)  subject to  ‖s‖ ≤ Δₖ
```
Comment (dpo) [inline src/R2N.jl:15]: Please update documentation.

File: src/R2N.jl [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    hk = h(xk[selected])
+    hk < Inf || error("prox computation must be erroneous")
+    verbose > 0 && @debug "R2N: found point where h has value" hk
+  end
+  hk == -Inf && error("nonsmooth term is not proper")
+
+  xkn = similar(xk)
+  s = zero(xk)
+  ψ =
+    (has_bounds(f) || subsolver == TRDH) ?
```
Comment (dpo) [inline src/R2N.jl:107]: Do you really want to use TRDH as subsolver? The point of R2N is to be able to compute a prox without a trust region. R2DH would be more appropriate.

File: src/R2N.jl [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+  Fobj_hist = zeros(maxIter)
+  Hobj_hist = zeros(maxIter)
+  Complex_hist = zeros(Int, maxIter)
+  if verbose > 0 
+    #! format: off
+    @info @sprintf "%6s %8s %8s %8s %7s %7s %8s %7s %7s %7s %7s %1s" "outer" "inner" "f(x)" "h(x)" "√ξ1" "√ξ" "ρ" "σ" "‖x‖" "‖s‖" "‖Bₖ‖" "R2N"
+    #! format: on
+  end
+
+    # main algorithm initialization
```
Comment (dpo) [inline src/R2N.jl:119]: indentation

File: src/R2N.jl [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+  k = 0
+
+  fk = obj(f, xk)
+  ∇fk = grad(f, xk)
+  ∇fk⁻ = copy(∇fk)
+
+  quasiNewtTest = isa(f, QuasiNewtonModel)
+  Bk = hess_op(f, xk)
+
+  λmax = (opnorm(Bk))
```
Comment (dpo) [inline src/R2N.jl:131]: extra parentheses

File: src/R2N.jl [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    ∇φ!(g, d) = begin
+      mul!(g, Bk, d)
+      g .+= ∇fk
+      g .+= σk * d
+      g
+    end
+
+    mk(d) = φ(d) + ψ(d)
+
+
```
Comment (dpo) [inline src/R2N.jl:159]: remove extra blank lines

File: src/R2N.jl [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+      set_bounds!(ψ, l_bound - xk, u_bound - xk)
+
+      #update functions
+      fk = fkn
+      hk = hkn
+
+      # update gradient & Hessian
+      shift!(ψ, xk)
+      ∇fk = grad(f, xk)
+      # grad!(f, xk, ∇fk)
```
Comment (dpo) [inline src/R2N.jl:229]: Why not use `grad!()` here?

File: src/R2N.jl [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+      ∇fk⁻ = ∇fk
+    end
+
+    if ρk < η1 || ρk == Inf
+        σk = σk * γ
+
+    end
+    
+    
+
```
Comment (dpo) [inline src/R2N.jl:246]: lots of extra blank lines

**Review by dpo — CHANGES_REQUESTED** [review]

**Comment by dpo** [comment]
@MohamedLaghdafHABIBOULLAH Please revise this PR and the one for R2DH.

File: src/R2N.jl [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+      ϵ_subsolver += ϵ_increment
+    end
+
+    if sqrt(ξ1) < ϵ
+      # the current xk is approximately first-order stationary
+      optimal = true
+      continue
+    end
+
+    if sub_tol
```
Comment (geoffroyleconte) [inline src/R2N.jl:176]: Thanks you for the PR! Can you explain what is `sub_tol` in the documentation of your function?

**Review by geoffroyleconte — CHANGES_REQUESTED** [review]

File: src/R2N.jl (hunk 8 above)
Comment (MohamedLaghdafHABIBOULLAH) [reply src/R2N.jl:176]: Ok

**Comment by dpo** [comment]
@MohamedLaghdafHABIBOULLAH Could you please rebase this PR?

@geoffroyleconte How does this look to you?

**Comment by geoffroyleconte** [comment]
LGTM, but I think we should wait https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl/pull/117 to incorporate its changes in your criticality measure.

**Comment by dpo** [comment]
@MohamedLaghdafHABIBOULLAH Please rebase this PR and make sure the comments above have been applied.

**Pull request closed without merging**